lib/wadler/print.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+# Oppen.
+module Oppen
+  # Wadler.
+  class Wadler
+    attr_reader :current_nest_indentation
+    attr_reader :margin
+    attr_reader :new_line
+    attr_reader :tokens
+
+    def initialize(margin: 80, new_line: "\n")
+      @current_nest_indentation = 0
+      @margin = margin
+      @new_line = new_line
+      @tokens = []
+    end
+
+    # @return [String]
+    def output
+      tokens << Oppen.eof
+      Oppen.print(tokens:, margin:, new_line:)
+    end
+
+    # @param indent [Integer] group indentation
+    # @param open_obj [String] group oppening delimiter
+    # @param close_obj [String] group closing delimiter
+    # @param break_type [Oppen::Token::BreakType] group breaking type
+    #
+    # @return [Nil]
+    def group(indent = 0, open_obj = '', close_obj = '',
+              break_type = Oppen::Token::BreakType::CONSISTENT, &)
+      if !open_obj.empty?
+        self.break
+        text(open_obj)
+      end
+
+      nest(indent, break_type, &)
+
+      if !close_obj.empty? # rubocop:disable Style/GuardClause
+        self.break
+        text(close_obj)
+      end
+    end
+
+    # @param indent [Integer] nest indentation
+    # @param break_type [Oppen::Token::BreakType] nest breaking type
+    #
+    # @return [Nil]
+    def nest(indent = 2, break_type = Oppen::Token::BreakType::CONSISTENT, &block)
+      last_nest_indentation = current_nest_indentation
+      @current_nest_indentation = indent
+
+      tokens << if break_type == Oppen::Token::BreakType::CONSISTENT
+        Oppen.begin_consistent(offset: indent)
+      else
+        Oppen.begin_inconsistent(offset: indent)
+      end
```

Comment:
That's wrong. We don't have a type system to enforce a type on `break_type`, so if you send a string, it will default to an inconsistent break. We want to fail.

We have 2 options: either use `case-in`:

```ruby
      tokens << case break_type
      in Oppen::Token::BreakType::CONSISTENT
        Oppen.begin_consistent(offset: indent)
      in Oppen::Token::BreakType::INCONSISTENT
        Oppen.begin_inconsistent(offset: indent)
      end
```

… which will be better formatted as such:

```ruby
      tokens << 
        case break_type
        in Oppen::Token::BreakType::CONSISTENT
          Oppen.begin_consistent(offset: indent)
        in Oppen::Token::BreakType::INCONSISTENT
          Oppen.begin_inconsistent(offset: indent)
        end
```

Another option we have is to use a [type checker](https://sorbet.org/) on top of our library.

I will be exploring this option soon. For the time being, let's use `case-in`.